src/libemulation/tools/Hex2bin-2.5/hex2bin.1
.\" Automatically generated by Pod::Man 2.28 (Pod::Simple 3.29)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{
.    if \nF \{
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "HEX2BIN 1"
.TH HEX2BIN 1 "2017-05-11" "perl v5.22.1" "User Contributed Perl Documentation"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
hex2bin/mot2bin  \e\- converts Intel/Motorola  hex files into binary
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
hex2bin [options] file
.PP
Option list:
    [\-a]                Address Alignment Word (hex2bin only)
    [\-b]                Batch mode: exits if specified file doesn't exist
    [\-c]                Enable record checksum verification
    [\-C Poly Init RefIn RefOut XorOut]
                        \s-1CRC\s0 parameters
    [\-e extension]      Output filename extension (without the dot)
    [\-E 0|1]            Endian for checksum/CRC, 0: little, 1: big
    [\-f address]        Address of check result to write
    [\-F address value]  Address and value to force
    [\-k 0\-5]            Select check method (checksum or \s-1CRC\s0) and size
    [\-d]                Display list of check methods/value size
    [\-l length]         Maximal Length (Starting address + Length \-1 is Max Address)
                        File will be filled with Pattern until Max Address is reached
    [\-m size]           Minimum Block Size
                        File Size Dimension will be a multiple of Minimum block size
                        File will be filled with Pattern
                        Length must be a power of 2 in hexadecimal [see \-l option]
                        Attention this option is \s-1STRONGER\s0 than Maximal Length
    [\-p pad byte]       Pad-byte value in hex (default: 0xFF)
    [\-r start end]      Range to compute checksum over (default is min and max addresses)
    [\-s address]        Starting address in hex for binary file (default: 0)
                         ex.: if the first record is :nn010000ddddd...
                        the data supposed to be stored at 0100 will start at 0000
                        in the binary file.
                        Specifying starting address with \-s 0100 will put pad bytes in the
                        binary file so that the data supposed to be stored at 0100
                        will start at the same address in the binary file.
    [\-t address]        Floor address in hex
    [\-T address]        Ceiling address in hex
    [\-v]                Verbose messages for debugging purposes
    [\-w]                Swap wordwise (low <\-> high)
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBHex2bin\fR
is a program that converts an Intel hex format into binary.
It can handle the extended Intel hex format. Both the segmented
and linear address records are supported.
Records need not be sorted and there can be gaps between records.
Records are assumed to be non-overlapping.
Padding bytes may be specified and a checksum may be inserted in the
binary file.
.PP
\&\fBMot2bin\fR
does the same with Motorola hex files. It has the same features and command line
options as hex2bin except where indicated. 24 bit and 32 bit records are supported.
.SH "OPTIONS"
.IX Header "OPTIONS"
Options can be specified in any order, with the file name at the end. Options are
now case sensitive. All option values are specified in hexadecimal.
.PP
\&\fB\-a\fR
.PP
Address Alignment Word. (hex2bin only)
.PP
Hex with record type, where data is represented in Word (2 Byte)
e.g Texas Instruments: \s-1TMS320F2835, TMS320F28065.\s0
.PP
\&\fB\-b\fR
.PP
Batch processing mode: the program won't ask for a file name if the one specified is wrong.
.PP
\&\fB\-c\fR
.PP
Enables checksum verification.
.PP
By default, it ignores checksum errors in the hex file, so that someone can change
by hand some bytes with a text editor, allowing quick fixes without recompiling a source
code all over again. This is useful when tweaking constants directly in the code or
something similar. If you want checksum error reporting, specify the option \-c.
.PP
\&\fBEx.: hex2bin \-c example.hex\fR
.PP
If there is a checksum error somewhere, the program will continue the
conversion anyway.
.PP
\&\fB\-C Poly Init RefIn RefOut XorOut\fR
.PP
\&\s-1CRC\s0 parameters. See the doc/CRC list.txt file for a description of common CRCs. See also
the test/Makefile for examples of command lines. Needs \fB\-k\fR and \fB\-f\fR option.
RefIn and RefOut parameters are specified by \fBt\fR or \fBf\fR for true or false.
.PP
\&\fB\-d\fR
.PP
Display the list of available check methods and sizes.
.PP
\&\fB\-e extension\fR
.PP
By default, the output file will have an extension \fBfilename.bin\fR.
Another extension may be specified with this command:
.PP
\&\fBEx.: hex2bin \-e com example.hex\fR
.PP
The output file will be example.com
.PP
\&\fB\-E 0|1\fR
.PP
Endianness for writing the check result or forcing a 16\-bit value.
    \fB0\fR: little, \fB1\fR: big.
.PP
By default, little endian is used.
.PP
\&\fB\-f address\fR
.PP
Address in hexadecimal for inserting the check value in the binary file. Needs \fB\-k\fR
option to specify the check method. A range can be specified with option \fB\-r\fR.
.PP
\&\fB\-F address value\fR
.PP
Address and value of checksum to insert (force) in the binary file. Needs \fB\-k\fR
option to specify the size. The value is written as is.
.PP
\&\fB\-k 0\-5\fR
.PP
In many cases, someone needs to insert a check value in the binary file. For example,
a boot rom is programmed with a checksum which is verified at power-up. This feature
uses also options \fB\-r\fR, \fB\-C\fR and \fB\-f\fR. Display the list of checksum/CRC with \fB\-d\fR.
.PP
Select the checksum type to insert into the binary file
 0:  Checksum  8\-bit
 1:  Checksum 16\-bit (adds 16\-bit words into a 16\-bit sum, data and result \s-1BE\s0 or \s-1LE\s0)
 2:  \s-1CRC8
 3:  CRC16
 4:  CRC32
 5: \s0 Checksum 16\-bit (adds bytes into a 16\-bit sum, result \s-1BE\s0 or \s-1LE\s0)
.PP
\&\fB\-l length\fR
.PP
The binary file will be padded with \s-1FF\s0 or pad bytes as specified by the option
below, up to a maximal Length (Starting address + Length \-1 is Max Address)
.PP
\&\fB\-m minimum_block_size\fR
.PP
File Size Dimension will be a multiple of Minimum block size.
File will be filled with Pattern.
Length must be a power of 2 in hexadecimal [see \fB\-l\fR option]
Attention this option is \s-1STRONGER\s0 than Maximal Length
.PP
\&\fB\-p pad_byte\fR
.PP
Pads unused locations with the specified byte.
.PP
By default, this byte is \s-1FF,\s0 which is the unprogrammed value for most EPROM/EEPROM/Flash.
.PP
\&\fBEx.: hex2bin \-p 3E example.hex\fR
.PP
\&\fB\-r [start] [end]\fR
.PP
Range to compute binary checksum over (default is min and max addresses)
.PP
\&\fB\-s address\fR
.PP
Specify the starting address of the binary file.
.PP
Normally, hex2bin will generate a binary file starting at the lowest address in
the hex file. If the lowest address isn't 0000, ex: 0100, the first byte that
should be at 0100 will be stored at address 0000 in the binary file. This may
cause problems when using the binary file to program an \s-1EPROM.\s0
.PP
If you can't specify the starting address (or offset) to your \s-1EPROM\s0 programmer,
you can specify a starting address on the command line:
.PP
\&\fBEx.: hex2bin \-s 0000 records_start_at_0100.hex\fR
.PP
The bytes will be stored in the binary file with a padding from 0000 to the
lowest address (00FF in this case). Padding bytes are all \s-1FF\s0 by default so an \s-1EPROM\s0
programmer can skip these bytes when programming. The padding value can be changed
with the \-p option.
.PP
\&\fB\-t address\fR
.PP
Specify the flooring address in hexadecimal of the binary file. (hex2bin only)
.PP
This option discards all lower addresses than the flooring address.
.PP
\&\fB\-T address\fR
.PP
Specify the ceiling address in hexadecimal of the binary file. (hex2bin only)
.PP
This option discards all higher addresses than the ceiling address.
\&\fB\-v\fR
.PP
Verbose mode for diagnostics
.PP
\&\fB\-w\fR
.PP
Swap wordwise (low <\-> high).
.SH "NOTES"
.IX Header "NOTES"
This program does minimal error checking since many hex files are
generated by known good assemblers.
.SH "AUTHOR Jacques Pelletier (jpelletier@ieee.org) \- version 2.5"
.IX Header "AUTHOR Jacques Pelletier (jpelletier@ieee.org) - version 2.5"


src/libemulation/tools/Hex2bin-2.5/hex2bin.c
/*
  hex2bin converts an Intel hex file to binary.

  Copyright (C) 2015,  Jacques Pelletier
  checksum extensions Copyright (C) 2004 Rockwell Automation
  All rights reserved.

  Redistribution and use in source and binary forms, with or without
  modification, are permitted provided that the following conditions are
  met:
  Redistributions of source code must retain the above copyright notice,
  this list of conditions and the following disclaimer.
  Redistributions in binary form must reproduce the above copyright notice,
  this list of conditions and the following disclaimer in the documentation
  and/or other materials provided with the distribution.
  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

  20040617 Alf Lacis: Added pad byte (may not always want FF).
  Added 'break;' to remove GNU compiler warning about label at
  end of compound statement
  Added PROGRAM & VERSION strings.

  20071005 PG: Improvements on options parsing
  20091212 JP: Corrected crash on 0 byte length data records
  20100402 JP: Corrected bug on physical address calculation for extended
  linear address record.
  ADDRESS_MASK is now calculated from MEMORY_SIZE

  20120125 Danny Schneider:
  Added code for filling a binary file to a given Max_Length relative to
  Starting Address if Max-Address is larger than Highest-Address
  20120509 Yoshimasa Nakane:
  modified error checking (also for output file, JP)
  20141005 JP: added support for byte swapped hex files
           corrected bug caused by extra LF at end or within file
  20141008 JP: removed junk code
  20141121 Slucx: added line for removing extra CR when entering file name at run time.
  20141122 Simone Fratini: small feature added
  20150116 Richard Genoud (Paratronic): correct buffer overflows/wrong results with the -l flag
  20150122 JP: added support for different check methods
  20150221 JP: rewrite of the checksum write/force value
  20150804 JP: added batch file option
  20160923 JP: added code for checking filename length
  20170418 Simone Fratini: added option -t and -T to obtain shorter binary files
*/

#define PROGRAM "hex2bin"
#define VERSION "2.5"

#include "common.h"

#define NO_ADDRESS_TYPE_SELECTED 0
#define LINEAR_ADDRESS 1
#define SEGMENTED_ADDRESS 2

const char *Pgm_Name = PROGRAM;

int main (int argc, char *argv[])
{
    /* line inputted from file */
    char Line[MAX_LINE_SIZE];

    /* flag that a file was read */
    bool Fileread;

    /* cmd-line parameter # */
    char *p;

    int result;

    /* Application specific */
    unsigned int First_Word, Address, Segment, Upper_Address;
    unsigned int Type;
    unsigned int Offset = 0x00;
    unsigned int temp;

    /* We will assume that when one type of addressing is selected, it will be valid for all the
     current file. Records for the other type will be ignored. */
    unsigned int Seg_Lin_Select = NO_ADDRESS_TYPE_SELECTED;

    unsigned int temp2;

    byte	Data_Str[MAX_LINE_SIZE];

    fprintf (stdout,PROGRAM" v"VERSION", Copyright (C) 2017 Jacques Pelletier & contributors\n\n");

    if (argc == 1)
        usage();

    strcpy(Extension, "bin"); /* default is for binary file extension */

	ParseOptions(argc, argv);

    /* when user enters input file name */

    /* Assume last parameter is filename */
    GetFilename(Filename,argv[argc -1]);

    /* Just a normal file name */
    NoFailOpenInputFile (Filename);
    PutExtension(Filename, Extension);
    NoFailOpenOutputFile(Filename);
    Fileread = true;

    /* When the hex file is opened, the program will read it in 2 passes.
    The first pass gets the highest and lowest addresses so that we can allocate
    the right size.
    The second pass processes the hex data. */

    /* To begin, assume the lowest address is at the end of the memory.
     While reading each records, subsequent addresses will lower this number.
     At the end of the input file, this value will be the lowest address.

     A similar assumption is made for highest address. It starts at the
     beginning of memory. While reading each records, subsequent addresses will raise this number.
     At the end of the input file, this value will be the highest address. */
    Lowest_Address = (unsigned int)-1;
    Highest_Address = 0;
    Records_Start = 0;
    Segment = 0;
    Upper_Address = 0;
    Record_Nb = 0;    // Used for reporting errors
    First_Word = 0;

    /* Check if are set Floor and Ceiling Address and range is coherent*/
    VerifyRangeFloorCeil();

    /* get highest and lowest addresses so that we can allocate the rintervallo incoerenteight size */
    do
    {
        unsigned int i;

        /* Read a line from input file. */
        GetLine(Line,Filin);
        Record_Nb++;

        /* Remove carriage return/line feed at the end of line. */
        i = strlen(Line);

        if (--i != 0)
        {
            if (Line[i] == '\n') Line[i] = '\0';

            /* Scan the first two bytes and nb of bytes.
               The two bytes are read in First_Word since its use depend on the
               record type: if it's an extended address record or a data record.
               */
            result = sscanf (Line, ":%2x%4x%2x%s",&Nb_Bytes,&First_Word,&Type,Data_Str);
            if (result != 4) fprintf(stderr,"Error in line %d of hex file\n", Record_Nb);

            p = (char *) Data_Str;

            /* If we're reading the last record, ignore it. */
            switch (Type)
            {
            /* Data record */
            case 0:
                if (Nb_Bytes == 0)
                    break;

                Address = First_Word;

                if (Seg_Lin_Select == SEGMENTED_ADDRESS)
                {
                    Phys_Addr = (Segment << 4) + Address;
                }
                else
                {
                    /* LINEAR_ADDRESS or NO_ADDRESS_TYPE_SELECTED
                       Upper_Address = 0 as specified in the Intel spec. until an extended address
                       record is read. */
                    Phys_Addr = ((Upper_Address << 16) + Address);
                }

                if (Verbose_Flag) fprintf(stderr,"Physical Address: %08X\n",Phys_Addr);

                /* Floor address */
                if (Floor_Address_Setted)
                {
                    /* Discard if lower than Floor_Address */
                  if (Phys_Addr < (Floor_Address - Starting_Address)) {
                    if (Verbose_Flag) fprintf(stderr,"Discard physical address less than %08X\n",Floor_Address - Starting_Address);
                    break; 
                  }
                }
                /* Set the lowest address as base pointer. */
                if (Phys_Addr < Lowest_Address)
                    Lowest_Address = Phys_Addr;

                /* Same for the top address. */
                temp = Phys_Addr + Nb_Bytes -1;

                /*Ceiling address */
                if (Ceiling_Address_Setted)
                {
                    /* Discard if higher than Ceiling_Address */
                    if (temp  > (Ceiling_Address +  Starting_Address)) {
                      if (Verbose_Flag) fprintf(stderr,"Discard physical address more than %08X\n",Ceiling_Address + Starting_Address);
                      break;
                    }
                }
                if (temp > Highest_Address) {
                    Highest_Address = temp;
                    if (Verbose_Flag) fprintf(stderr,"Highest_Address: %08X\n",Highest_Address);
                }
                break;


            case 1:
                if (Verbose_Flag) fprintf(stderr,"End of File record\n");
                break;

            case 2:
                /* First_Word contains the offset. It's supposed to be 0000 so
                   we ignore it. */

                /* First extended segment address record ? */
                if (Seg_Lin_Select == NO_ADDRESS_TYPE_SELECTED)
                    Seg_Lin_Select = SEGMENTED_ADDRESS;

                /* Then ignore subsequent extended linear address records */
                if (Seg_Lin_Select == SEGMENTED_ADDRESS)
                {
                    result = sscanf (p, "%4x%2x",&Segment,&temp2);
					if (result != 2) fprintf(stderr,"Error in line %d of hex file\n", Record_Nb);

                    if (Verbose_Flag) fprintf(stderr,"Extended Segment Address record: %04X\n",Segment);

                    /* Update the current address. */
                    Phys_Addr = (Segment << 4);
                }
                else
                {
                    fprintf(stderr,"Ignored extended linear address record %d\n", Record_Nb);
                }
                break;

            case 3:
                if (Verbose_Flag) fprintf(stderr,"Start Segment Address record: ignored\n");
                break;

            case 4:
                /* First_Word contains the offset. It's supposed to be 0000 so
                   we ignore it. */

                /* First extended linear address record ? */
                if (Seg_Lin_Select == NO_ADDRESS_TYPE_SELECTED)
                    Seg_Lin_Select = LINEAR_ADDRESS;

                /* Then ignore subsequent extended segment address records */
                if (Seg_Lin_Select == LINEAR_ADDRESS)
                {
                    result = sscanf (p, "%4x%2x",&Upper_Address,&temp2);
					if (result != 2) fprintf(stderr,"Error in line %d of hex file\n", Record_Nb);

                    if (Verbose_Flag) fprintf(stderr,"Extended Linear Address record: %04X\n",Upper_Address);

                    /* Update the current address. */
                    Phys_Addr = (Upper_Address << 16);

                    if (Verbose_Flag) fprintf(stderr,"Physical Address: %08X\n",Phys_Addr);
                }
                else
                {
                    fprintf(stderr,"Ignored extended segment address record %d\n", Record_Nb);
                }
                break;

            case 5:
                if (Verbose_Flag) fprintf(stderr,"Start Linear Address record: ignored\n");
                break;

            default:
                if (Verbose_Flag) fprintf(stderr,"Unknown record type: %d at %d\n",Type,Record_Nb);
                break;
            }
        }
    }
    while (!feof (Filin));

    if (Address_Alignment_Word)
        Highest_Address += (Highest_Address - Lowest_Address) + 1;

    Allocate_Memory_And_Rewind();

    Segment = 0;
    Upper_Address = 0;
    Record_Nb = 0;

    /* Read the file & process the lines. */
    do /* repeat until EOF(Filin) */
    {
        unsigned int i;

        /* Read a line from input file. */
        GetLine(Line,Filin);
        Record_Nb++;

        /* Remove carriage return/line feed at the end of line. */
        i = strlen(Line);

        //fprintf(stderr,"Record: %d; length: %d\n", Record_Nb, i);

        if (--i != 0)
        {
            if (Line[i] == '\n') Line[i] = '\0';

            /* Scan the first two bytes and nb of bytes.
               The two bytes are read in First_Word since its use depend on the
               record type: if it's an extended address record or a data record.
            */
            result = sscanf (Line, ":%2x%4x%2x%s",&Nb_Bytes,&First_Word,&Type,Data_Str);
            if (result != 4) fprintf(stderr,"Error in line %d of hex file\n", Record_Nb);

            Checksum = Nb_Bytes + (First_Word >> 8) + (First_Word & 0xFF) + Type;

            p = (char *) Data_Str;

            /* If we're reading the last record, ignore it. */
            switch (Type)
            {
            /* Data record */
            case 0:
                if (Nb_Bytes == 0)
                {
                    fprintf(stderr,"0 byte length Data record ignored\n");
                    break;
                }

                Address = First_Word;

                if (Seg_Lin_Select == SEGMENTED_ADDRESS)
                    Phys_Addr = (Segment << 4) + Address;
                else
                    /* LINEAR_ADDRESS or NO_ADDRESS_TYPE_SELECTED
                       Upper_Address = 0 as specified in the Intel spec. until an extended address
                       record is read. */
                    if (Address_Alignment_Word)
                        Phys_Addr = ((Upper_Address << 16) + (Address << 1)) + Offset;
                    else
                        Phys_Addr = ((Upper_Address << 16) + Address);

                /* Check that the physical address stays in the buffer's range. */
                if ((Phys_Addr >= Lowest_Address) && (Phys_Addr <= Highest_Address))
                {
                    /* The memory block begins at Lowest_Address */
                    Phys_Addr -= Lowest_Address;

					p = ReadDataBytes(p);

                    /* Read the Checksum value. */
                    result = sscanf (p, "%2x",&temp2);
                    if (result != 1) fprintf(stderr,"Error in line %d of hex file\n", Record_Nb);

					/* Verify Checksum value. */
                    Checksum = (Checksum + temp2) & 0xFF;
                    VerifyChecksumValue();
                }
                else
                {
                    if (Seg_Lin_Select == SEGMENTED_ADDRESS)
                        fprintf(stderr,"Data record skipped at %4X:%4X\n",Segment,Address);
                    else
                        fprintf(stderr,"Data record skipped at %8X\n",Phys_Addr);
                }

                break;

            /* End of file record */
            case 1:
                /* Simply ignore checksum errors in this line. */
                break;

            /* Extended segment address record */
            case 2:
                /* First_Word contains the offset. It's supposed to be 0000 so
                   we ignore it. */

                /* First extended segment address record ? */
                if (Seg_Lin_Select == NO_ADDRESS_TYPE_SELECTED)
                    Seg_Lin_Select = SEGMENTED_ADDRESS;

                /* Then ignore subsequent extended linear address records */
                if (Seg_Lin_Select == SEGMENTED_ADDRESS)
                {
                    result = sscanf (p, "%4x%2x",&Segment,&temp2);
                    if (result != 2) fprintf(stderr,"Error in line %d of hex file\n", Record_Nb);

                    /* Update the current address. */
                    Phys_Addr = (Segment << 4);

                    /* Verify Checksum value. */
                    Checksum = (Checksum + (Segment >> 8) + (Segment & 0xFF) + temp2) & 0xFF;
					VerifyChecksumValue();
                }
                break;

            /* Start segment address record */
            case 3:
                /* Nothing to be done since it's for specifying the starting address for
                   execution of the binary code */
                break;

            /* Extended linear address record */
            case 4:
                /* First_Word contains the offset. It's supposed to be 0000 so
                   we ignore it. */

                if (Address_Alignment_Word)
                {
                    sscanf (p, "%4x",&Offset);
                    Offset = Offset << 16;
                    Offset -= Lowest_Address;

                }
                /* First extended linear address record ? */
                if (Seg_Lin_Select == NO_ADDRESS_TYPE_SELECTED)
                    Seg_Lin_Select = LINEAR_ADDRESS;

                /* Then ignore subsequent extended segment address records */
                if (Seg_Lin_Select == LINEAR_ADDRESS)
                {
                    result = sscanf (p, "%4x%2x",&Upper_Address,&temp2);
                    if (result != 2) fprintf(stderr,"Error in line %d of hex file\n", Record_Nb);

                    /* Update the current address. */
                    Phys_Addr = (Upper_Address << 16);

                    /* Verify Checksum value. */
                    Checksum = (Checksum + (Upper_Address >> 8) + (Upper_Address & 0xFF) + temp2)
                               & 0xFF;
					VerifyChecksumValue();
                }
                break;

            /* Start linear address record */
            case 5:
                /* Nothing to be done since it's for specifying the starting address for
                   execution of the binary code */
                break;
            default:
                fprintf(stderr,"Unknown record type\n");
                break;
            }
        }
    }
    while (!feof (Filin));
    /*-----------------------------------------------------------------------------*/

    fprintf(stdout,"Binary file start = %08X\n",Lowest_Address);
    fprintf(stdout,"Records start     = %08X\n",Records_Start);
    fprintf(stdout,"Highest address   = %08X\n",Highest_Address);
    fprintf(stdout,"Pad Byte          = %X\n",  Pad_Byte);

	WriteMemory();

#ifdef USE_FILE_BUFFERS
    free (FilinBuf);
    free (FiloutBuf);
#endif

    fclose (Filin);
    fclose (Filout);

    if (Status_Checksum_Error && Enable_Checksum_Error)
    {
        fprintf(stderr,"Checksum error detected.\n");
        return 1;
    }

    if (!Fileread)
        usage();
    return 0;
}
